### PR TITLE
JP-3153: post-commissioning AMI3 updates

### PR DESCRIPTION
**Pull request by rcooper295**
Resolves JP-3153 and the following sub-tasks and linked tickets: 
- [~JP-1713~](https://jira.stsci.edu/browse/JP-1713)
- [~JP-1714~](https://jira.stsci.edu/browse/JP-1714)
- [JP-2904](https://jira.stsci.edu/browse/JP-2904)
- [JP-3346](https://jira.stsci.edu/browse/JP-3346)
- [JP-3348](https://jira.stsci.edu/browse/JP-3348)

This PR addresses all changes to the NIRISS AMI mode analysis software since its last delivery in 2020. The main change is a complete overhaul of output formats. Two new datamodels were created: AmiOIModel (https://github.com/spacetelescope/stdatamodels/pull/174) and AmiLgFitModel (https://github.com/spacetelescope/stdatamodels/pull/199).

**Step changes:**
ami_analyze: 
 - Additional input arguments and defaults: affine2d, src, bandpass, usebp, firstfew, chooseholes, run_bpfix. 
 - Outputs are now two types of OIFITS files: averaged across integrations and per-integration, and the AmiLgFitModel which stores the cropped/cleaned data, model, and residuals. The OIFITS format is a standard in the field for storing interferometric observables; the AmiLgFitModel product is mostly intended as a diagnostic tool. 
 - The filter throughput reference file used previously was too coarsely sampled across the bandpass; now WebbPSF filter definitions are used and combined with a source spectrum from the Phoenix library (A0V by default) using `synphot` and `stsynphot`. 
 - A median filter is used to identify integrations with bright CR hits not identified by the jump detection step (e.g. when NGROUPS <=3) and DO_NOT_USE flags added to the DQ array. 
 - A fourier-based bad pixel correction is applied by default to fix pixels flagged DO_NOT_USE and/or JUMP_DET. 

ami_average:
- This step will be turned off in the default AMI3 pipeline flow until the effects of averaging together multiple reference star observations are better understood. The code is left as-is except "skip = True" has been added to the spec block. This step will not work if run manually on the AmiOIModel products.

ami_normalize:
- Rewritten to use the averaged AmiOIModel datamodels.

Other changes include processing 3D instead of 2D images by default for AMI exposures (rateints and calints files)
The calwebb_ami3 pipeline has been updated to run only the ami_analyze and ami_normalize steps. **Association generator needs to be updated**

**Checklist for maintainers**
- [ ] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [ ] added relevant milestone
- [ ] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by braingram** [comment]
FWIW this roman PR adds fetching a webpsf data file to the CI (which is similar to the CI here in jwst):
https://github.com/spacetelescope/romancal/pull/794/files#diff-fe43a405829f0a62513f19fdcc051dcbba563c0ca9b2d16eacf543ca3a9e52deR50
I believe @zacharyburnett helped @bmorris3 make those additions.

**Comment by rcooper295** [comment]
I did notice the docs build was failing because of something having to do with the MASK_NRM file it's trying to get from WebbPSF. We are also now getting filter throughput files from WebbPSF too. Those files will be required with any run of the AMI3 steps; the mask file is required when the bad pixel correction is run, which is the default. It's currently fetching the WEBBPSF_PATH env variable. Is that still an OK way to do it if we add checking that environment variable as part of the CI as in the Roman case you shared? Is that in the tox.ini file?

File: jwst/ami/bp_fix.py
Comment (braingram) [inline jwst/ami/bp_fix.py:49]:
```suggestion
pupil_masks = {}
webbpsf_path = os.getenv('WEBBPSF_PATH')
if webbpsf_path is not None:
    pupilfile_nrm = os.path.join(webbpsf_path,'NIRISS/optics/MASK_NRM.fits.gz')
    pupil_masks["NRM"] = fits.getdata(pupilfile_nrm)
else:
    # TODO maybe a warning here that WEBBPSF_PATH is not defined?
    # warnings.warn("WEBBPSF_PATH environment variable is undefined...")
```

I wonder if something like this might fix the docs build. This would leave `pupil_masks["NRM"]` undefined when `WEBBPSF_PATH` is not defined so some of the other code in this PR will almost certainly fail but I'm not sure if the docs require any of that to run.

**Comment by braingram** [comment]
I added a suggestion to hopefully make the docs build.

For the CI (running on github) and the regtests (run on jenkins) a few additional changes will be needed. I'm not sure the exact changes that will be required but I think both `.github/workflows/ci.yml` (and `ci_cron.yml`) will need to be updated to fix the github actions and `JenkinsfileRT` and `JenkinsRT_dev` updated to fix the regtests. Let me see if I can open a PR against your branch to test edits to at least the github actions.

File: jwst/ami/bp_fix.py
Comment (rcooper295) [reply jwst/ami/bp_fix.py:49]: I can make that change so hopefully the docs build would work, though I think it would have the same issue when it encounters the similar lines in ami/utils.py: https://github.com/rcooper295/jwst/blob/amianalyze/jwst/ami/utils.py#L1220
I could also have it raise an exception in the beginning ami_analyze_step.py where I check some of the arguments if WEBBPSF_PATH isn't in the environment, since the code will break without it?

Comment (braingram) [reply jwst/ami/bp_fix.py]: The docs look almost built :) The issue now appears to be a missing `throughput_reffile` file which is referenced in `docs/jwst/references_general/references_general.rst`. Should there be a `throughput_reffile` in the folder with that file?

Comment (rcooper295) [reply jwst/ami/bp_fix.py]: Getting there, one error at a time... No, I removed the use of the throughput ref file as part of the changes to the step (it used too few wavelengths across the bandpass to be useful). I'll remove the reference to the throughput file in the docs there and see if that fixes it. I have a few more documentation changes to make still that hopefully won't break it again.

Comment (braingram) [reply jwst/ami/bp_fix.py]: Unfortunately I know nothing about that file or if it should be removed from crds.

The good news is... the docs built!

**Comment by braingram** [comment]
https://github.com/spacetelescope/jwst/pull/7873 has the changes in this PR and updates to:
- `.github/workflows/ci.yml` (this appears functional but not optimal due to needing to fetch the webbpsf data for every run)
- `JenkinsfileRT` (these are as yet untested, the server is running other tests right now)
I opened a PR against your branch (see: https://github.com/rcooper295/jwst/pull/2) with these changes. Once `ci.yml` is sorted the weekly `ci_cron.yml` will also need to be updated and the same for `JenkinsfileRT_dev` (for the devdeps regression tests).

File: jwst/ami/oifits.py
Comment (braingram) [inline jwst/ami/oifits.py:258]: Can some (or all) of this be read from the existing meta data (or defaults/nans filled) instead of making a Simbad query? This would allow dropping the astroquery dependency and I'm not sure making queries to external services will work well in operations?

File: jwst/ami/utils.py
Comment (braingram) [inline jwst/ami/utils.py:1222]: Can this be replaced with updated "throughput" reference files on CRDS?

File: jwst/ami/bp_fix.py
Comment (braingram) [inline jwst/ami/bp_fix.py:46]: Is there a file with this data available through CRDS? If not, can this be added as a new reference file?

File: jwst/ami/oifits.py
Comment (rcooper295) [reply jwst/ami/oifits.py:258]: It probably can be replaced; I'll look into this.

File: jwst/ami/utils.py
Comment (braingram) [inline jwst/ami/utils.py:1322]: Is there a function in poppy that can be used here? It looks like it's already a dependency of the pipeline.

File: jwst/ami/oifits.py
Comment (braingram) [inline jwst/ami/oifits.py:239]: The current truth files are failing oifits validation (using the oifits-check from [oifitslib](https://github.com/jsy1001/oifitslib)).

Here are the errors for `jw01093012001_03102_00001_nis_a3002_ami-oi.fits`:
```
*** Does not conform to the OIFITS standard:
Reference to missing array element, 27 occurrences:-
    OI_VIS #1 record 1
    OI_VIS #1 record 2
    OI_VIS #1 record 3
    OI_VIS #1 record 4
    OI_VIS #1 record 5
    OI_VIS #1 record 6
    OI_VIS2 #1 record 1
    OI_VIS2 #1 record 2
    OI_VIS2 #1 record 3
    [List truncated]
*** Does not conform to the OIFITS standard:
OI_T3 table may contain unnormalised triple product amplitude, 24 occurrences:-
    OI_T3 #1 record 3 channel 1
    OI_T3 #1 record 5 channel 1
    OI_T3 #1 record 6 channel 1
    OI_T3 #1 record 7 channel 1
    OI_T3 #1 record 8 channel 1
    OI_T3 #1 record 9 channel 1
    OI_T3 #1 record 10 channel 1
    OI_T3 #1 record 11 channel 1
    OI_T3 #1 record 13 channel 1
    [List truncated]
```

The errors might be related to this line where `sta_index` is computed. The `OI_ARRAY` table in the file does not contain a line for `sta_index` 0. However the first 6 rows of `OI_VIS` and `OI_VIS2` list `sta_index` 0 as does many rows in `OI_T3`.

File: jwst/ami/oifits.py
Comment (braingram) [inline jwst/ami/oifits.py:52]: I'd like to improve the readability of these variables. I'm assuming `nbl` is `n_baselines`. Is this correct? Also, what is `ncp` and `nca`? This comment applies to many variables throughout this PR.

File: jwst/ami/utils.py
Comment (rcooper295) [reply jwst/ami/utils.py]: Thanks, done in recent commit!

File: jwst/ami/oifits.py
Comment (rcooper295) [reply jwst/ami/oifits.py:52]: I think there is a bit of a history to the short variable names here; one of the original authors learned to code in C on punch cards where character counts were stricter, and is now not comfortable with a lot of typing for physical reasons. Nevertheless I can see the advantage of more descriptive variable names; I'm a bit afraid of breaking things by doing large-scale renaming at this point though...

File: jwst/ami/oifits.py
Comment (rcooper295) [reply jwst/ami/oifits.py:239]: I missed a bit of reformatting; seems that the OIFITS standard wants these indices to be 1-indexed rather than 0-indexed. Just fixed that!

File: jwst/ami/oifits.py
Comment (braingram) [reply jwst/ami/oifits.py:52]: Thanks for the context.

I think the renaming will be easier to do as part of this PR (as the code is "fresh" in mind and testing is ongoing) rather than at some later date. It will help for later maintenance, debugging, and modifications. I must admit that I am not an expert when it comes to the algorithms used here but I also expect the same is true for some of the folks that will maintain this code.

I am happy to help with these changes with some guidance on what names to replace. Would it be helpful if I added more comments to variable names that are candidates for renaming?

Comment (rcooper295) [reply jwst/ami/oifits.py:52]: If your coding environment has a more reliable way of updating all usages of a given variable name than doing find/replace in each file that would be helpful; I'm not sure how to do that efficiently in the editor I'm using. 
In the lines you highlighted above, nbl is indeed n_baselines. nh = n_holes, ncp = n_closurephases, nca = n_closureamplitudes, fp = fringephases, fa = fringeamplitudes. Those are the ones that I saw immediately as good candidates for renaming, but if you find others that it would be helpful to have my input on let me know!
As always, thanks for your help in making this code a bit more readable!

File: jwst/ami/nrm_core.py
Comment (emolter) [inline jwst/ami/nrm_core.py:64]: I know this wasn't introduced by this PR, but this seems to me a strange way to use **kwargs.  The long chain of if...else statements could be eliminated by listing each of the three keyword arguments as optional arguments to the function with default values.

File: jwst/ami/nrm_core.py
Comment (emolter) [inline jwst/ami/nrm_core.py:227]: Move this long comment block into docstring, perhaps under Notes section

File: jwst/ami/oifits.py
Comment (emolter) [inline jwst/ami/oifits.py:23]: can this be referenced in some way?  Also, should probably be moved to Notes

File: jwst/ami/oifits.py
Comment (emolter) [inline jwst/ami/oifits.py:78]: add a description for this function

File: jwst/ami/tests/test_ami_interface.py
Comment (emolter) [inline jwst/ami/tests/test_ami_interface.py:36]: Is it possible to also add a few random NaNs or large values here to add test coverage for all the bad pixel replacement, or is that too naive?

**Comment by braingram** [comment]
Regression tests started using (non-test) `crds` (I will update this comment with results):
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1293/

For this run some failures are expected as the updates in this PR (including the new regression tests) will have issues with `jwst-crds` not yet having the `nrm` reference type.

I will start another run (and update this comment) with the new regression tests and `crds-test`.

File: jwst/ami/utils.py
Comment (emolter) [inline jwst/ami/utils.py:1373]: this line about implaneia can probably be removed

File: jwst/ami/utils.py
Comment (emolter) [inline jwst/ami/utils.py:1401]: should this use log.warning instead?

**Review by emolter — COMMENTED** [review]
This looks fantastic to me overall, thanks so much @rcooper295 and everyone else on the team who contributed to this!  I'm pretty unfamiliar with the underlying math, so most of my comments are just attempting to improve clarity of the docs and code.  I'm still pretty new to this, so feel free to ignore anything you feel is incorrect or unnecessary to change.

**Comment by rcooper295** [comment]
Thanks so much for your review @emolter! I am going through the suggested changes and will mark each of your comments as resolved as I cover them.

File: jwst/ami/bp_fix.py
Comment (rcooper295) [reply jwst/ami/bp_fix.py]: I removed the Short Summary headers in this file, but they are ubiquitous in this code; maybe they were the recommended style when the previous version was developed. Should I go through and remove them everywhere they occur?

Comment (emolter) [reply jwst/ami/bp_fix.py]: I don't feel like that decision is up to me.  Maybe @hbushouse or @tapastro can chime in

File: jwst/ami/instrument_data.py
Comment (rcooper295) [reply jwst/ami/instrument_data.py:227]: These lines just create a list of duplicates of the bandpass array equal to the number of integrations being processed, so it's not necessary when there's only a single integration. Not entirely sure why the code was set up like this though.

**Comment by rcooper295** [comment]
I think I have addressed all of the comments except those dealing with the tests, which are the domain of @braingram I think (though happy to help with those if I can)!

File: jwst/ami/bp_fix.py
Comment (braingram) [reply jwst/ami/bp_fix.py:236]: The regtests hit the lines you mentioned (confirmed by setting breakpoints and running them locally). I'll add comments about the `example_model` to your other comment.

File: jwst/ami/tests/test_ami_interface.py
Comment (braingram) [reply jwst/ami/tests/test_ami_interface.py:36]: That's a great question. Unfortunately my response is... maybe.

`example_model` here is rather unrepresentative of real data. I tried setting a `nan` value but there are a few issues.

One limitation at the moment is the lack of access to the `nrm` reference file on `crds` (it only exists on `crds-test` which the unit tests don't have access to). This would cause all the unit tests to fail, so I added a mock version ([above the example_model fixture](https://github.com/spacetelescope/jwst/pull/7862/files#diff-25ac33e148190671f89968855ac6fd4c6507a758584aaba2efba1c07e4691b35R13)) which at the moment creates a `NRMModel` with an empty (all zero) mask. This is also unrepresentative of the real reference file (but allows the step added in this PR to run).

I'm not sure if it's the faked mask or the faked data but if I add a `nan` value:
```
model.data[0, 30, 30] = np.nan
```
The `test_ami_analyze_step` fails with:
```
        # The fractional part of the offsets:
>       d0, d1 = (c[0] - int(c[0]), c[1] - int(c[1]))
E       ValueError: cannot convert float NaN to integer

jwst/ami/hextransformee.py:114: ValueError
```
If instead I use
```
model.data[:, 30, 30] = np.nan
```
The test passes and the nan-check in `bp_fix.find_bad_pixels` is covered but `bp_fix.fourier_corr` is not (due to the blank mask).

@rcooper295 is there a "minimal" model here of the mask and input data that you expect should pass with some bad pixel flagging?

File: jwst/ami/bp_fix.py
Comment (tapastro) [reply jwst/ami/bp_fix.py]: I would vote to do this if possible, as it doesn't match our existing style. Sorry for the hassle!

**Comment by tapastro** [comment]
Looking through the review and comments, I see one outstanding discussion on testing bad pixel replacement. Is that the current status of this PR? Is there anything else that needs to be done? Hoping this can be merged in advance of our deadline so we can get testing started early on this large change. I'll start a read-through as well.

**Comment by rcooper295** [comment]
I think that's pretty much the status -- @braingram opened this PR to address some of the testing issues https://github.com/rcooper295/jwst/pull/7. I'll go through the code today and remove those pesky "Short Summary" docstring headers. It would be great to get it merged as soon as possible to allow everyone downstream enough time to do their testing but I don't want to inadvertently introduce a bug by being too hasty.

**Comment by braingram** [comment]
@tapastro one additional issue is the `nrm` reference file availability on `crds-test` only. Is there an eta for when this will be available on `crds`? In the meantime should the regtests that require this (the new ones in this PR) be xfailed?

File: jwst/ami/leastsqnrm.py
Comment (braingram) [reply jwst/ami/leastsqnrm.py]: Good catch! Unit test added in: https://github.com/spacetelescope/jwst/pull/7862/commits/f982ebfc446c7e7552124f1dd12e5debcb1c2e94#diff-efd78172dc65690ea634c2cce2d6b91291ecf67f0b5a91a968272dd0d59c7b1bR6

File: jwst/ami/nrm_core.py
Comment (braingram) [reply jwst/ami/nrm_core.py]: This was refactored in https://github.com/spacetelescope/jwst/pull/7862/commits/f982ebfc446c7e7552124f1dd12e5debcb1c2e94

**Comment by tapastro** [comment]
> @tapastro one additional issue is the `nrm` reference file availability on `crds-test` only. Is there an eta for when this will be available on `crds`? In the meantime should the regtests that require this (the new ones in this PR) be xfailed?

Our goal is to have the CRDS-OPS server updated this week, prior to the build deadline. If the CRDS status is the last holdup for this PR, it's in a good spot. I'll update this PR when I hear more about the CRDS server upgrade progress.

**Comment by hbushouse** [comment]
@rcooper295 The PR is showing a couple of file conflicts with the latest updates. Can you try to resolve those so that we can get a final test run of this before merging?

**Comment by hbushouse** [comment]
FYI: I started a full regtest run using this PR branch and the latest CRDS-TEST context (which contains NRM and all other routine updates) at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1350/

**Comment by hbushouse** [comment]
Minor failure in style check due to import of `scipy.integrate.simpson` no longer being used (it was deprecated in another recent PR).

**Comment by hbushouse** [comment]
Latest regtest run https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1350/ using the latest CRDS-TEST context shows failures that are only due to different ref file version numbers between CRDS-OPS and CRDS-TEST. No differences in actual data results. So this looks good.

**Comment by braingram** [comment]
@tapastro @emolter any further comments on this PR or is it ready for approval and merge?

**Comment by tapastro** [comment]
> @tapastro @emolter any further comments on this PR or is it ready for approval and merge?

The only holdup I would have is build-related - I don't know the status of that conversation. I would defer to Howard, but I don't have anything else to hold up this from being merged.

**Comment by hbushouse** [comment]
I think it's good to go.

**Review by hbushouse — APPROVED** [review]

File: CHANGES.rst
Comment (braingram) [reply CHANGES.rst:7]: I re-added that entry in https://github.com/spacetelescope/jwst/pull/7862/commits/04aecaa9d14d664975a3a1dbe76319e37d3dfe43

**Comment by emolter** [comment]
> @tapastro @emolter any further comments on this PR or is it ready for approval and merge?

All my comments are addressed, so it looks good to me.  However, when I looked at this a few hours ago, there were some unit test failures I didn't understand.  I don't think they had anything to do with the AMI step, so as long as someone knows what's causing those I'm happy

**Comment by hbushouse** [comment]
The CI failures are almost certainly due to the use of stcal 1.6.1, which was missing some recent api changes to the jump step. Running against stcal 1.7.0 should fix those.

**Pull request merged**
